contracts/bridge/contracts/binance/BasicOutboundLane.sol

```diff
@@ -0,0 +1,293 @@
+        verifyMMRLeaf(beefyMMRLeaf, beefyMMRLeafIndex, beefyMMRLeafCount, peaks, siblings);
+        verifyMessages(
+            subOutboundLaneDataHash,
+            hash(subInboundLaneData),
+            beefyMMRLeaf,
+            chainCount,
+            chainMessagesProof,
+            laneMessagesRoot,
+            laneCount,
+            laneMessagesProof
+        );
+        DeliveredMessages memory confirmed_messages = confirm_delivery(subInboundLaneData);
+        // TODO: callback `on_messages_delivered`
+        pay_relayers_rewards(subInboundLaneData.relayers, confirmed_messages.begin, confirmed_messages.end);
+    }
+
+    /* Private Functions */
+
+    function extract_substrate_inbound_lane_info(InboundLaneData memory lane_data) internal pure returns (uint256 total_unrewarded_messages, uint256 last_delivered_nonce) {
+        uint256 len = lane_data.relayers.length;
+        if(len > 0) {
+            UnrewardedRelayer memory front = lane_data.relayers[0];
+            UnrewardedRelayer memory back = lane_data.relayers[len-1];
+            total_unrewarded_messages = back.messages.end - front.messages.begin + 1;
+            last_delivered_nonce = back.messages.end;
+        } else {
+            total_unrewarded_messages = 0;
+            last_delivered_nonce = lane_data.last_confirmed_nonce;
+        }
+    }
+
+    function confirm_delivery(InboundLaneData memory subInboundLaneData) internal returns (DeliveredMessages memory confirmed_messages) {
+        (uint256 total_messages, uint256 latest_delivered_nonce) = extract_substrate_inbound_lane_info(subInboundLaneData);
+        require(total_messages < 256, "Lane: InvalidNumberOfMessages");
+
+        UnrewardedRelayer[] memory relayers = subInboundLaneData.relayers;
+        OutboundLaneData memory dataMem = data;
+        require(latest_delivered_nonce > dataMem.latest_received_nonce, "Lane: NoNewConfirmations");
+        require(latest_delivered_nonce <= dataMem.latest_generated_nonce, "Lane: FailedToConfirmFutureMessages");
+        require(latest_delivered_nonce - dataMem.latest_received_nonce <= total_messages, "Lane: TryingToConfirmMoreMessagesThanExpected");
+        uint256 dispatch_results = extract_dispatch_results(dataMem.latest_received_nonce, latest_delivered_nonce, relayers);
+        uint256 prev_latest_received_nonce = dataMem.latest_received_nonce;
+        data.latest_received_nonce = latest_delivered_nonce;
+        confirmed_messages = DeliveredMessages({
+            begin: prev_latest_received_nonce + 1,
+            end: latest_delivered_nonce,
+            dispatch_results: dispatch_results
+        });
+        emit MessagesDelivered(lanePosition, confirmed_messages.begin, confirmed_messages.end, confirmed_messages.dispatch_results);
+    }
+
+    function extract_dispatch_results(uint256 prev_latest_received_nonce, uint256 latest_received_nonce, UnrewardedRelayer[] memory relayers) internal pure returns(uint256 received_dispatch_result) {
```

Comment:
this might be simple if using struct similar to following?

```
#[derive(Clone, Default, Encode, Decode, RuntimeDebug, PartialEq, Eq)]
pub struct DeliveredMessages {
	/// Nonce of the first message that has been delivered (inclusive).
	pub begin: MessageNonce,
	/// Nonce of the last message that has been delivered (inclusive).
	pub end: MessageNonce,
	/// Dispatch result (`false`/`true`), returned by the message dispatcher for every
	/// message in the `[begin; end]` range. See `dispatch_result` field of the
	/// `bp_runtime::messages::MessageDispatchResult` structure for more information.
	pub dispatch_results: DispatchResultsBitVec,
}
```

Reply:
I am indeed using this data structure here, except for the optimization of using uint256 for `dispatch_results`.